src/libcrun/utils.c

```diff
@@ -1915,36 +1935,61 @@ find_executable (const char *executable_path, const char *cwd)
   if (executable_path[0] == '/')
     {
       ret = check_access (executable_path);
-      if (ret == 0)
-        return xstrdup (executable_path);
-      return NULL;
+      if (LIKELY (ret == 0))
+        {
+          *out = xstrdup (executable_path);
+          return 0;
+        }
+      last_error = ret;
+      goto fail;
     }
 
   end = tmp = xstrdup (getenv ("PATH"));
 
   while ((it = strsep (&end, ":")))
     {
-      size_t len;
+      cleanup_free char *path = NULL;
 
       if (it == end)
         it = ".";
 
-      len = snprintf (path, PATH_MAX, "%s/%s", it, executable_path);
-      if (len >= PATH_MAX)
-        continue;
+      ret = append_paths (&path, err, it, executable_path, NULL);
+      if (UNLIKELY (ret < 0))
+        {
+          crun_error_release (err);
+          continue;
+        }
 
       ret = check_access (path);
       if (ret == 0)
-        return xstrdup (path);
+        {
+          /* Change owner.  */
+          *out = path;
+          path = NULL;
+          return 0;
+        }
 
-      if (errno == ENOENT)
+      if (ret == -ENOENT)
         continue;
 
-      last_error = errno;
+      last_error = ret;
     }
 
-  errno = last_error;
-  return NULL;
+fail:
+  switch (last_error)
+    {
+    case CHECK_ACCESS_NOT_EXECUTABLE:
+      return crun_make_error (err, EPERM, "the path `%s` exists but it is not executable", executable_path);
+
+    case CHECK_ACCESS_NOT_REGULAR:
+      return crun_make_error (err, EPERM, "the path `%s` is not a regular file", executable_path);
+
+    default:
+      errno = -last_error;
+      if (errno == ENOENT)
+        return crun_make_error (err, errno, "executable file `%s` not found in $PATH", executable_path);
```

Comment:
I think this needs an extra `executable_path[0] != '/'` check here before adding `in $PATH` because with that you also add it for `/blah` which was not looked up in $PATH of course so the error is misleading.
Maybe something like `executable_path[0] != '/' ? " in $PATH" : ""` and then `%s` in message